File: 2017/day4/day4.py
from typing import List, Optional, Tuple


class Passphrase:
    def getInput(self) -> List[List[str]]:
        inputFile = "input-test.txt" if self.useTest else "input.txt"
        with open(inputFile, "r") as file1:
            return [line.strip().split(" ") for line in file1]

    def __init__(self, useTest: Optional[bool] = False) -> None:
        self.useTest = useTest
        self.stringList = self.getInput()

    def getValid(self, isPart2: Optional[bool] = False) -> int:
        count = 0
        for strings in self.stringList:
            wordDict = set()
            for word in strings:
                key = self.generateKey(word) if isPart2 else word
                if key in wordDict:
                    break
                wordDict.add(key)
            else:
                count += 1
        return count

    def generateKey(self, word: str) -> Tuple[str]:
        charList = [0] * 26
        for char in word:
            charList[ord(char) - ord("a")] += 1
        return tuple(charList)


if __name__ == "__main__":
    passphrase = Passphrase()
    print("Day 4 part 1:", passphrase.getValid())
    print("Day 4 part 2:", passphrase.getValid(True))

File: 2019/day4/day4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Container struct {
	UseTest bool
	start   int
	end     int
}

func (this *Container) getInput() {
	inputFile := "input.txt"
	if this.UseTest {
		inputFile = "input-test.txt"
	}
	file, _ := os.Open(inputFile)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		startStr, endStr := strings.Split(line, "-")[0], strings.Split(line, "-")[1]
		this.start, _ = strconv.Atoi(startStr)
		this.end, _ = strconv.Atoi(endStr)
	}
	defer file.Close()
}
func (w *Container) checkInput(n int, isPart2 bool) bool {
	strNum := strconv.Itoa(n)
	prevDigit, _ := strconv.Atoi(string(strNum[0]))
	sameCounter := 0
	same := false

	for _, char := range strNum[1:] {
		digit, _ := strconv.Atoi(string(char))

		if digit < prevDigit {
			return false
		}

		if digit == prevDigit {
			sameCounter++
		} else {
			if (isPart2 && sameCounter == 1) || (!isPart2 && sameCounter >= 1) {
				same = true
			}
			sameCounter = 0
		}
		prevDigit = digit
	}
	return (isPart2 && (same || sameCounter == 1)) || (!isPart2 && (same || sameCounter >= 1))
}

func (this *Container) countPasswords(isPart2 bool) int {
	start := max(100_000, this.start)
	end := min(999_999, this.end)
	count := 0
	for i := start; i <= end; i++ {
		if this.checkInput(i, isPart2) {
			count++
		}
	}
	return count
}

func main() {
	container := &Container{
		UseTest: false,
	}
	container.getInput()
	fmt.Println("Day 3 part 1:", container.countPasswords(false))
	fmt.Println("Day 3 part 1:", container.countPasswords(true))

}
